### PR TITLE
Aggregate functions for graph processing

### PR DESCRIPTION
**Pull request by Hattonuri**
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 Aggregate functions for graph processing

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by Hattonuri** [comment]
### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
 Aggregate functions for graph processing


> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

File: src/Storages/HDFS/StorageHDFS.h [hunk 1]
```diff
@@ -133,8 +133,8 @@ class HDFSSource : public SourceWithProgress, WithContext
     Block block_for_format;
     std::vector<NameAndTypePair> requested_virtual_columns;
     UInt64 max_block_size;
-    bool need_path_column;
-    bool need_file_column;
+    [[maybe_unused]] bool need_path_column;
```
Comment (Hattonuri) [inline src/Storages/HDFS/StorageHDFS.h:136]: Без этого не компилится

**Review by CurtizJ — COMMENTED** [review]
Please add:
1. Some tests with distributed queries to check whether `serialize/deserialize/merge` works. You can use table function `remote('127.0.0.{1,2}', database, table)` for that.
2. Comments about algorithms you are using to implement operations with graphs.

File: src/AggregateFunctions/registerGraphAggregateFunctions.cpp [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+BOOST_PP_SEQ_FOR_EACH(EXTERN_GRAPH_OPERATION_FACTORY, , GRAPH_FUNCTION_CLASSES)
+
+void registerGraphAggregateFunctions(AggregateFunctionFactory & factory)
+{
+#define REGISTER_GRAPH_FUNCTION(r, data, elem) \
+    factory.registerFunction(BOOST_PP_STRINGIZE(elem), {BOOST_PP_CAT(createGraphOperation, elem), AggregateFunctionProperties{}});
+
+    BOOST_PP_SEQ_FOR_EACH(REGISTER_GRAPH_FUNCTION, , GRAPH_FUNCTION_CLASSES)
+}
```
[the remark this answers is not in the document]
Comment (Hattonuri) [reply src/AggregateFunctions/registerGraphAggregateFunctions.cpp:29]: We can't make it good:
TreeDiameter has template parameter - so we can't extern it here. Error about unknown template will be thrown(This solution looked like it until I made optimization for simple classes)
We could have included all aggregate functions classes here with making changing from .cpp to .h - but we would have registerGraphAggregateFunction compiling too much time and on chaging any aggregate function. With increasing the number of aggregate function the problem would increase
This is "probably" not really good-looking solution, but it's really easy to maintain: on adding new operation you should only add it's classname to GRAPH_FUNCTION_CLASSES. Also it has advantage that it contains usings - so you will have all hints without using this->

File: src/AggregateFunctions/AggregateFunctionGraphBidirectionalData.h [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        }
+        else
+        {
+            const char * begin = nullptr;
+            key = columns[0]->serializeValueIntoArena(row_num, *arena, begin);
+            value = columns[1]->serializeValueIntoArena(row_num, *arena, begin);
+        }
+        graph[key].push_back(value);
+        graph[value].push_back(key);
+        ++edges_count;
```
[the remark this answers is not in the document]
Comment (Hattonuri) [reply src/AggregateFunctions/AggregateFunctionGraphBidirectionalData.h:42]: Yes, we should - other algorithms do, in the comment below you've found the issue

**Pull request closed without merging**